quinn-proto/src/connection/send_buffer.rs

```diff
@@ -31,18 +33,46 @@ impl SendBuffer {
 
     /// Append application data to the end of the stream
     pub fn write(&mut self, data: &[u8]) {
-        self.unacked.extend_from_slice(data);
+        let buf: Bytes = Bytes::from(data.to_owned());
+        self.unacked_segments.push_back(buf);
+        self.unacked_len += data.len();
         self.offset += data.len() as u64;
     }
 
     /// Discard a range of acknowledged stream data
     ///
     /// Each offset must be acknowledged at most once.
-    pub fn ack(&mut self, range: Range<u64>) {
+    pub fn ack(&mut self, mut range: Range<u64>) {
+        // Clamp the range to data which is still tracked
+        let base_offset = self.offset - self.unacked_len as u64;
+        range.start = base_offset.max(range.start);
+        range.end = base_offset.max(range.end);
```

Comment:
I believe this is an unrelated change; it should be split off into its own commit (and the doc comment updated, if we're relaxing the constraint).